libethereum/EthereumPeer.cpp

```diff
@@ -220,8 +220,13 @@ void EthereumPeer::tick()
 	auto s = session();
 	time_t  now = std::chrono::system_clock::to_time_t(chrono::system_clock::now());
 	if (s && (now - m_lastAsk > 10 && m_asking != Asking::Nothing))
+	{
 		// timeout
 		s->disconnect(PingTimeout);
+
+		if (m_observer)
+			m_observer->onPeerRequestTimeout(dynamic_pointer_cast<EthereumPeer>(shared_from_this()), m_asking);
```

Comment:
Because enable_shared_from_this is in the base class (Capability), so shared_from_this() here gives us shared_ptr<Capabilty>
It is this way when handling all responses in this class